trace/http.go

```diff
@@ -2,19 +2,46 @@ package trace
 
 import "net/http"
 
-func Get(url string, parent *http.Request) (resp *http.Response, err error) {
-	c := &http.Client{}
-	req, err := http.NewRequest("GET", url, nil)
-	if err != nil {
-		return nil, err
+// DefaultCollector is a MemoryCollector. Trace defaults back to DefaultCollector if none given.
+var DefaultCollector = NewMemoryCollector()
+
+// Trace is used to add incoming Trace functionality.
+type Trace struct {
+	collector Collector
+}
+
+func (t *Trace) process(rw http.ResponseWriter, r *http.Request, f http.HandlerFunc) {
+	if t.collector == nil {
+		t.collector = DefaultCollector
 	}
-	SetIdIfMissing(req, parent)
-	return c.Do(req)
+
+	span := NewSpanIDFromRequest(r)
+	event := NewEvent(span, RequestReceived)
+	t.collector.Record(event)
+
+	r2 := SetHeaders(r, span)
+	f(rw, r2)
+
+	event = NewEvent(span, RequestCompleted)
+	t.collector.Record(event)
+}
+
+// Handler wraps an existing http.Handler with Trace functionality.
+func (t *Trace) Handler(h http.Handler) http.Handler {
+	f := h.ServeHTTP
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		t.process(rw, r, f)
+	})
 }
 
-func Instrument(fn http.HandlerFunc) http.HandlerFunc {
+// HandlerFunc wraps an existing http.HandlerFunc with Trace functionality.
+func (t *Trace) HandlerFunc(f http.HandlerFunc) http.HandlerFunc {
```

Comment:
I think this is unnecessary and the `Handler` function should be enough. You would rarely wrap a function.